src/meta/processors/job/BalanceJobExecutor.cpp

```diff
@@ -130,371 +113,164 @@ nebula::cpp2::ErrorCode BalanceJobExecutor::save(const std::string& k, const std
-    int32_t hostsSize = zoneHosts_[zoneName].size();
-    int32_t totalPartsZone = 0;
-    for (auto& host : zoneHosts_[zoneName]) {
-      auto it = confirmedHostParts.find(host);
-      if (it == confirmedHostParts.end()) {
-        LOG(ERROR) << "Host " << host << "not in confirmedHostParts";
-        continue;
+    avgPartNum = totalPartNum / hostVec.size();
+    size_t remainder = totalPartNum - avgPartNum * hostVec.size();
+    size_t leftBegin = 0;
+    size_t leftEnd = 0;
+    size_t rightBegin = 0;
+    size_t rightEnd = hostVec.size();
+    std::vector<BalanceTask> tasks;
+    for (size_t i = 0; i < hostVec.size(); i++) {
+      if (avgPartNum <= hostVec[i]->parts_.size()) {
+        leftEnd = i;
+        break;
       }
-      totalPartsZone += it->second.size();
     }
-
-    avgLoad = static_cast<float>(totalPartsZone) / hostsSize;
-    minLoad = std::floor(avgLoad);
-    maxLoad = std::ceil(avgLoad);
-    LOG(INFO) << "Update min and max loading Total parts in zone " << totalPartsZone
-              << ", total hosts " << hostsSize << " The expect avg load is " << avgLoad
-              << " The min load is " << minLoad << " max load is " << maxLoad;
-  }
-
-  while (maxPartsHost.second > maxLoad || minPartsHost.second < minLoad) {
-    auto& partsFrom = confirmedHostParts[maxPartsHost.first];
-    auto& partsTo = confirmedHostParts[minPartsHost.first];
-    std::sort(partsFrom.begin(), partsFrom.end());
-    std::sort(partsTo.begin(), partsTo.end());
-
-    LOG(INFO) << maxPartsHost.first << ":" << partsFrom.size() << " -> " << minPartsHost.first
-              << ":" << partsTo.size();
-    std::vector<PartitionID> diff;
-    std::set_difference(partsFrom.begin(),
-                        partsFrom.end(),
-                        partsTo.begin(),
-                        partsTo.end(),
-                        std::inserter(diff, diff.begin()));
-    bool noAction = true;
-    for (auto& partId : diff) {
-      LOG(INFO) << "partsFrom size " << partsFrom.size() << " partsTo size " << partsTo.size()
-                << " minLoad " << minLoad << " maxLoad " << maxLoad;
-      if (partsFrom.size() == partsTo.size() + 1 ||
-          partsFrom.size() == static_cast<size_t>(minLoad) ||
-          partsTo.size() == static_cast<size_t>(maxLoad)) {
-        VLOG(3) << "No need to move any parts from " << maxPartsHost.first << " to "
-                << minPartsHost.first;
+    for (size_t i = 0; i < hostVec.size(); i++) {
+      if (avgPartNum < hostVec[i]->parts_.size()) {
+        rightBegin = i;
         break;
       }
-
-      LOG(INFO) << "[space:" << spaceId << ", part:" << partId << "] " << maxPartsHost.first << "->"
-                << minPartsHost.first;
-      auto it = std::find(partsFrom.begin(), partsFrom.end(), partId);
-      if (it == partsFrom.end()) {
-        LOG(ERROR) << "Part " << partId << " not found in partsFrom";
-        return false;
+    }
+    for (size_t right = rightBegin; right < rightEnd;) {
```

Comment:
We have two layer for loop here:
```
for (size_t right = rightBegin; right < rightEnd;) {
  for (; leftIndex < leftEnd - 1; leftIndex++) {
  }
}
```

But I think only one is enough? Every move of a partition in `balance in zone` is legal.

Reply:
> We have two layer for loop here:
> 
> ```
> for (size_t right = rightBegin; right < rightEnd;) {
>   for (; leftIndex < leftEnd - 1; leftIndex++) {
>   }
> }
> ```
> 
> But I think only one is enough? Every move of a partition in `balance in zone` is legal.

the loop "for (; leftIndex < leftEnd - 1; leftIndex++)" is doing a Bubble Sort